lisa/node.py

```diff
@@ -270,47 +278,64 @@ def close(self) -> None:
-        node = None
-        if node_type is None:
-            raise LisaException("type of node shouldn't be None")
-        if node_type in [
-            constants.ENVIRONMENTS_NODES_LOCAL,
-            constants.ENVIRONMENTS_NODES_REMOTE,
-        ]:
-            node = Node.create(
-                len(self._list), node_type=node_type, is_default=node_runbook.is_default
-            )
-            self._list.append(node)
-            if node.is_remote:
-                fields = [
-                    constants.ENVIRONMENTS_NODES_REMOTE_ADDRESS,
-                    constants.ENVIRONMENTS_NODES_REMOTE_PORT,
-                    constants.ENVIRONMENTS_NODES_REMOTE_PUBLIC_ADDRESS,
-                    constants.ENVIRONMENTS_NODES_REMOTE_PUBLIC_PORT,
-                    constants.ENVIRONMENTS_NODES_REMOTE_USERNAME,
-                    constants.ENVIRONMENTS_NODES_REMOTE_PASSWORD,
-                    constants.ENVIRONMENTS_NODES_REMOTE_PRIVATE_KEY_FILE,
-                ]
-                parameters: Dict[str, Any] = dict()
-                for field in fields:
-                    value = getattr(node_runbook, field)
-                    if value is not None:
-                        parameters[field] = value
-                node.set_connection_info(**parameters)
+    def from_local(self, node_runbook: schema.LocalNode) -> Node:
+        assert isinstance(
+            node_runbook, schema.LocalNode
+        ), f"actual: {type(node_runbook)}"
+        node = Node.create(
+            len(self._list),
+            node_type=node_runbook.type,
+            is_default=node_runbook.is_default,
+        )
+        self._list.append(node)
+
         return node
 
-    def from_spec(
-        self,
-        spec: schema.NodeSpec,
-        node_type: str = constants.ENVIRONMENTS_NODES_REMOTE,
-    ) -> Node:
+    def from_remote(self, node_runbook: schema.RemoteNode) -> Optional[Node]:
+        assert isinstance(
+            node_runbook, schema.RemoteNode
+        ), f"actual: {type(node_runbook)}"
+
         node = Node.create(
-            len(self._list), spec=spec, node_type=node_type, is_default=spec.is_default
+            len(self._list),
+            node_type=node_runbook.type,
+            is_default=node_runbook.is_default,
         )
         self._list.append(node)
+
+        fields = [
+            constants.ENVIRONMENTS_NODES_REMOTE_ADDRESS,
+            constants.ENVIRONMENTS_NODES_REMOTE_PORT,
+            constants.ENVIRONMENTS_NODES_REMOTE_PUBLIC_ADDRESS,
+            constants.ENVIRONMENTS_NODES_REMOTE_PUBLIC_PORT,
+            constants.ENVIRONMENTS_NODES_REMOTE_USERNAME,
+            constants.ENVIRONMENTS_NODES_REMOTE_PASSWORD,
+            constants.ENVIRONMENTS_NODES_REMOTE_PRIVATE_KEY_FILE,
+        ]
+        parameters: Dict[str, Any] = dict()
+        for field in fields:
+            value = getattr(node_runbook, field)
+            if value is not None:
+                parameters[field] = value
```

Comment:
I've seen this pattern of "update this dict with this other dict, but only for these keys" being repeated in the code base.  Maybe it's time to refactor it?

Maybe something like:

```python

def filter_dict_fields(src_dict: Dict[str, Any], fields) -> Dict[str, Any]:
    fields = set(fields)
    return {k: v for k, v in src_dict.items() if k in fields and v is not None}
```

And use like so:

```python
parameters: Dict[str, Any] = filter_dict_fields(node_runbook, fields)
```

What do you think?

Reply:
Added a funciton like below

def set_filter_fields(src: Any, dest: Any, fields: List[str]) -> None:
    """
    copy field values form src to dest, if it's not None
    """
    assert src
    assert dest
    assert fields
    for field_name in fields:
        if hasattr(src, field_name):
            field_value = getattr(src, field_name)
        else:
            raise LisaException("field doesn't exist on src")
        if field_value is not None:
            setattr(dest, field_name, field_value)